### PR TITLE
feat: Support for PKCS10 and refactor of import/export formats

### PR DESCRIPTION
**Pull request by bgrieder**
### Features

- Support for certificate generation using the Certify KMIP operation and a PKCS#10 or a public key
- Support for most standardized encoding formats on import.export: PKCS#8, PKCS#1, SEC1, X509, PKCS#12
- Improvements to the Locate functionality for attributes and tags
- Support for the Get Attributes KMIP operation
- Database: support for more elaborate atomic operations
- Replaced part of Rust Crypto with openssl for more standardized module support
- Deactivated automatic verification, which will be reallocated for the future Validate KMIP operation support
- Deactivated certificate "quick create feature," which will be revisited to better fit in KMIP specs

### Bug Fixes

- Redis-Findex: `create` now checks for the pre-existence of the object
- Better KMIP compliance:
  - for KeyBlock [#76](https://github.com/Cosmian/kms/issues/76)
  - default export formats [#78](https://github.com/Cosmian/kms/issues/78)

## This is a BIG PR
Many files are affected by the `UniqueIdentifier` fix and the import/export format fix.

The only files really worth checking for errors are the server-side files:

In `crate/server/src/core/operations/`
- `certify.rs`
- `export.rs`
- `import.rs`

In `crate/server/src/core/certificate/`
- `find.rs`
- `tags.rs`

The rest of the modifications should really be covered by the unit tests and the clippy linter.

### CODE REVIEW

File: crate/server/src/core/operations/certify.rs [hunk 1]
```diff
@@ -12,30 +37,121 @@ use crate::{
...
+            "Certificate Request Type CRMF not supported".to_string()
+        )),
+        None => kms_bail!(KmsError::InvalidRequest(
+            "the certificate Request Type must be provided (PEM or PKCS10)".to_string()
+        )),
+    }?;
+
+    // Create an X509 struct with the desired certificate information.
+    let mut x509_builder = X509::builder().unwrap();
+    x509_builder.set_version(3)?;
```
Comment (JosePisco) [inline crate/server/src/core/operations/certify.rs:118]: X509 version is 0-indexed in OpenSSL. Setting version 3 should be
```suggestion
    x509_builder.set_version(2)?;
```
but I think defining a constant like
```
pub const X509_VERSION3: i32 = 2;
```
in `crate/server/src/core/certificate/mod.rs` gives more readability in the code ?

File: Cargo.toml [hunk 2]
```diff
@@ -1,12 +1,12 @@
 [workspace]
 members = [
-  "crate/cli",
-  "crate/client",
-  "crate/kmip",
-  "crate/server",
-  "crate/utils",
-  "crate/pyo3",
-  "crate/logger",
+    "crate/cli",
```
Comment (ThibsG) [inline Cargo.toml:3]: can we stick to old indent please?

File: Cargo.toml [hunk 3]
```diff
@@ -37,41 +37,48 @@ base64 = "0.21"
...
 cloudproof = { version = "2.3", features = ["findex-redis"] }
+const-oid = { version = "0.9", features = ["db"] }
+der = { version = "0.7", features = ["pem"] }
+env_logger = "0.10"
 hex = "0.4"
 http = "0.2"
 josekit = "0.8.3"
 native-tls = "0.2"
+num-bigint = { version = "0.4", features = ["rand", "serde"] }
 num-bigint-dig = { version = "0.8", default-features = false, features = [
```
Comment (ThibsG) [inline Cargo.toml:57]: we shouldn't be using `num-bigint` anymore right? keep only `num-bigint-dig` instead?

File: Cargo.toml [hunk 4]
```diff
@@ -37,41 +37,48 @@ base64 = "0.21"
...
-  "json",
-  "multipart",
-  "native-tls",
-  "rustls-tls",
-  "stream",
+# When using client certificate authentication, reqwest will use the
+# native-tls crate to create an Identity; this will be different backend
+# implementations depending on the operating system. On linux, it will
+# use a vendored version of openssl
+reqwest = { version = "0.11.14", features = [
```
Comment (ThibsG) [inline Cargo.toml:76]:
```suggestion
reqwest = { version = "0.11.22", features = [
```

File: crate/cli/Cargo.toml [hunk 5]
```diff
@@ -22,9 +22,12 @@ cosmian_kmip = { path = "../kmip" }
...
 cosmian_kms_utils = { path = "../utils" }
 cosmian_logger = { path = "../logger" }
+der = { workspace = true }
 hex = { workspace = true }
-oauth2 = "4.4"
-openssl = { workspace = true }
+## openssl is only used to parse PKCS#12 files because native support is not yet available in Rust yet.
+## It should not be used for anything else and will be removed once native support is available.
+openssl = { version = "0.10", features = ["vendored"] }
+oauth2 = { version = "4.4" }
```
Comment (ThibsG) [inline crate/cli/Cargo.toml:30]:
```suggestion
oauth2 = "4.4"
```

File: crate/cli/Cargo.toml [hunk 6]
```diff
@@ -39,16 +42,19 @@ tokio = { workspace = true }
...
 assert_cmd = "2.0"
+const-oid = { workspace = true }
 cosmian_kms_client = { path = "../client", default-features = false }
 cosmian_kms_server = { path = "../server", features = [
     "insecure",
 ], default-features = false }
+oauth2 = { version = "4.4" }
 predicates = "3.0"
-regex = { version = "1.10", default-features = false }
+regex = { version = "1.7", default-features = false }
```
Comment (ThibsG) [inline crate/cli/Cargo.toml:59]:
```suggestion
regex = { version = "1.10", default-features = false }
```

File: crate/cli/Cargo.toml [hunk 7]
```diff
@@ -39,16 +42,19 @@ tokio = { workspace = true }
...
 [dev-dependencies]
 actix-rt = { workspace = true }
 actix-server = { version = "2.3", default-features = false }
 assert_cmd = "2.0"
+const-oid = { workspace = true }
 cosmian_kms_client = { path = "../client", default-features = false }
 cosmian_kms_server = { path = "../server", features = [
     "insecure",
 ], default-features = false }
+oauth2 = { version = "4.4" }
```
Comment (ThibsG) [inline crate/cli/Cargo.toml:57]: already in `[dependencies]`
```suggestion
```

File: crate/client/Cargo.toml [hunk 8]
```diff
@@ -23,3 +23,4 @@ tokio-util = { workspace = true }
 url = { workspace = true }
 rustls = { workspace = true }
 webpki-roots = { workspace = true }
+log = "0.4.17"
```
Comment (ThibsG) [inline crate/client/Cargo.toml:26]: unused
```suggestion
```

File: crate/server/Cargo.toml [hunk 9]
```diff
@@ -34,11 +34,12 @@ alcoholic_jwt = "4091"
...
 async-trait = { workspace = true }
 base64 = { workspace = true }
-chrono = { workspace = true }
 clap = { workspace = true }
 cloudproof = { workspace = true }
-cosmian_kmip = { path = "../kmip" }
+cosmian_kmip = { path = "../kmip", features = ["openssl"] }
 cosmian_kms_utils = { path = "../utils", features = ["curve25519"] }
+ratls = { workspace = true }
+tee_attestation = { workspace = true }
```
Comment (ThibsG) [inline crate/server/Cargo.toml:42]: keep alphabetical order please (move back these 2 lines right after `openssl`)

File: crate/server/Cargo.toml [hunk 10]
```diff
@@ -74,6 +73,7 @@ url = { workspace = true }
 uuid = { workspace = true }
 x509-parser = { workspace = true }
 zeroize = { workspace = true }
+chrono = "0.4.31"
```
Comment (ThibsG) [inline crate/server/Cargo.toml:76]:
```rust
chrono = { workspace = true }
```
and check for alphabetical order

File: crate/utils/src/crypto/certificate/mod.rs [hunk 11]
```diff
@@ -1 +1 @@
-pub mod attributes;
+//pub mod attributes;
```
Comment (JosePisco) [inline crate/utils/src/crypto/certificate/mod.rs:1]: Might also want to delete certificate directory at this point ?

File: crate/utils/src/crypto/wrap/encrypt_decrypt.rs [hunk 12]
```diff
@@ -100,6 +98,27 @@ where
     }
 }
 
+fn encrypt_with_public_key(
+    wrapping_key: PKey<Public>,
+    plaintext: &[u8],
+) -> Result<Vec<u8>, KmipUtilsError> {
+    // wrap using ECIES/RSA
+    let encrypt_system = HybridEncryptionSystem::new("public_key_uid", wrapping_key);
+    let request = Encrypt {
```
Comment (JosePisco) [inline crate/utils/src/crypto/wrap/encrypt_decrypt.rs:107]: Why build an Encrypt request object only to feed it with the plaintext if the method on which it is called only use the plaintext field of the object ?

Same goes for the decrypt version of this method where a Decrypt object is created.

**Review by JosePisco — COMMENTED** [review]
Nothing to say regarding cryptographic usage.
Refer to my comments for changes.

File: crate/server/src/core/certificate/tags.rs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        .and_then(|cn| cn.data().as_utf8().ok())
+    {
+        let cn_tag = format!("_cert_cn={subject_common_name}");
+        user_tags.insert(cn_tag);
+    }
+    Ok(())
+}
+
+/// Get the `SubjectKeyIdentifier` X509 extension value
+/// If it not available, it is
```
Comment (ThibsG) [inline crate/server/src/core/certificate/tags.rs:39]:
```suggestion
/// If it is not available, it is
```

File: crate/server/src/core/implementation.rs [hunk 14]
```diff
@@ -118,24 +121,28 @@ impl KMS {
...
                             Ok(Box::new(AesGcmSystem::instantiate(&owm.id, &owm.object)?)
                                 as Box<dyn EncryptionSystem>)
                         }
                         other => {
-                            kms_not_supported!("symmetric encryption with algorithm: {other:?}")
+                            kms_not_supported!(
+                                "symmetric encryption with algorithm: {}",
+                                other
+                                    .map(|alg| alg.to_string())
+                                    .unwrap_or("[N/A]".to_string())
```
Comment (ThibsG) [inline crate/server/src/core/implementation.rs:144]:
```suggestion
                                other.map_or("[N/A]".to_string(), |alg| alg.to_string())
```

File: crate/server/src/core/implementation.rs [hunk 15]
```diff
@@ -209,38 +205,29 @@ impl KMS {
...
+                        Some(CryptographicAlgorithm::AES) => {
                             Ok(Box::new(AesGcmSystem::instantiate(&owm.id, &owm.object)?))
                         }
                         other => {
-                            kms_not_supported!("symmetric decryption with algorithm: {other:?}")
+                            kms_not_supported!(
+                                "symmetric decryption with algorithm: {}",
+                                other
+                                    .map(|alg| alg.to_string())
+                                    .unwrap_or("[N/A]".to_string())
```
Comment (ThibsG) [inline crate/server/src/core/implementation.rs:229]:
```suggestion
                                other.map_or("[N/A]".to_string(), |alg| alg.to_string())
```

File: crate/server/src/core/operations/certify.rs (hunk 1 above)
Comment (Manuthor) [reply crate/server/src/core/operations/certify.rs]: done in https://github.com/Cosmian/kms/commit/5eff99f9be8c2226cc47fb8644d88257554bf093

File: crate/utils/src/crypto/certificate/mod.rs (hunk 11 above)
Comment (Manuthor) [reply crate/utils/src/crypto/certificate/mod.rs]: done in 2b1a3c262d233c02f86172ada0a02f0df3ef1fbf

**Pull request merged**